src/databricks/labs/ucx/assessment/crawlers.py

```diff
@@ -250,3 +269,150 @@ def snapshot(self) -> list[ClusterInfo]:
+        return sha256(bytes(hash_value_string.encode("utf-8"))).hexdigest()
+
+    def _get_cluster_from_task(
+        self, task: RunTask, job_run: BaseRun, all_clusters: dict[str, ClusterDetails]
+    ) -> ClusterSpec | ClusterDetails:
+        """
+        Determine the cluster associated with the task
+        1) Look for new_cluster on the task
+        2) Look for existing cluster on the task
+        3) Look for cluster_instance on the task (filled when the task is run)
+        4) Look for cluster instance on the job (filled when the job is run)
+        """
+        if task.new_cluster is not None:
+            return task.new_cluster
+        elif task.existing_cluster_id is not None:
+            # from api docs If existing_cluster_id, the ID of an existing cluster that is used for all runs of this job.
+            job_clusters: dict[str, JobCluster] = {x.job_cluster_key: x for x in job_run.job_clusters}
+            return job_clusters[task.existing_cluster_id].new_cluster
+        elif task.cluster_instance is not None:
+            # fall back option 1
+            return all_clusters[task.cluster_instance.cluster_id]
+        elif job_run.cluster_instance is not None:
+            # fall back option 2
+            return all_clusters[job_run.cluster_instance.cluster_id]
+
+    def _get_spark_version_from_task(
+        self, task: RunTask, job_run: BaseRun, all_clusters: dict[str, ClusterDetails]
+    ) -> str:
+        """
+        Returns the spark version of the task cluster
+        """
+        return self._get_cluster_from_task(task, job_run, all_clusters).spark_version
+
+    def _get_data_security_mode_from_task(
+        self, task: RunTask, job_run: BaseRun, all_clusters: dict[str, ClusterDetails]
+    ) -> str:
+        """
+        Returns the security mode of the task cluster
+        """
+        data_security_mode = self._get_cluster_from_task(task, job_run, all_clusters).data_security_mode
+        return data_security_mode.value if data_security_mode is not None else None
+
+    def _assess_job_runs(
+        self, all_clusters: dict[str, ClusterDetails], all_job_runs: list[BaseRun], all_jobs: list[BaseJob]
+    ) -> list[ExternallyOrchestratedJobTask]:
+        """
+        Returns a list of ExternallyOrchestratedJobs
+        """
+        ext_orc_job_tasks: list[ExternallyOrchestratedJobTask] = list[ExternallyOrchestratedJobTask]()
+        all_persisted_job_ids = [x.job_id for x in all_jobs]
+        not_persisted_job_runs = list(
+            filter(lambda jr: jr.job_id is None or jr.job_id not in all_persisted_job_ids, all_job_runs)
+        )
+        for job_run in not_persisted_job_runs:
+            for task in job_run.tasks:
+                spark_version = self._get_spark_version_from_task(task, job_run, all_clusters)
+                data_security_mode = self._get_data_security_mode_from_task(task, job_run, all_clusters)
+                hashed_id = self._create_hash_from_job_run_task(task)
```

Comment:
I figured we can do dedupe in the actual report based on the hash.

Airflow creates job runs yes, but each task is triggered in a workflow, so a job run might exist that only has one task out of 5 that ran successfully, also each task can have it's own cluster configuration, and that's what we're checking.

Reply:
two things:
- this table has to get as little noise as possible, so we have to dedupe before we write to table.
- we have to do dedupe based on the job_run, not the task, because that's how multi-task jobs are designed.

Reply:
Yes, but cluster definitions live on the task as mentioned, trying to imagine a user running this..

Potentially we run the risk, in particularly large multi-task workflows, of directing the user to the wrong task to address.

If there are multiple tasks that would fail, then a user has to potentially change multiple cluster configurations. If we roll this up to one result, then we run the risk of telling the user to fix "1" thing.

Reply:
For job with 50 tasks we have to have only one record in this table.

Reply:
The challenge is around usability, if this shows up in failure JSON, returning a useful report from this is hard.

Reply:
Especially if there's 50 tasks ☝️

Reply:
I'll change the grain for now so we can move this along.

Reply:
But we'll have to make a decision about which cluster version info we surface.

A multi task cluster can use multiple cluster definitions.

For now I'll just choose the lowest dbr, but this can be misleading.